### PR TITLE
Engaging Crowds: enable indexed subject selection

### LINKED ISSUE #1876: Classifier: Enable subject selection by subject ID in the subjects store
**Package**
lib-classifier

**Is your feature request related to a problem? Please describe.**
When a subject ID is passed into the classifier as a prop, that subject should be loaded into the store and shown as the first classification subject.

When a workflow has prioritised (sequential) selection enabled, subsequent subjects in the queue should follow on from the selected subject eg. if I select page 15 from a volume, I'd expect pages 16, 17 etc. to follow in the queue.

**Describe the solution you'd like**
I'll open an issue on Panoptes for the server-side component for this, but we could pass the subject ID when making a request to `/api/subjects/queued`.

Alternatively, we could wait until the specified subject has been classified before requesting the rest of the queue. I think that should work, since prioritised selection ought to send back subjects based on the last subject that you classified.

### PR DESCRIPTION
**Pull request by eatyourgreens**
Get indexed subject selection working for Engaging Crowds projects. Indexed selection shows you every subject from an indexed set in order. ~~Setting this to draft because I think `subjects.populateQueue()` might still need some refactoring.~~ The logic for subject selection was getting messy, so I've moved the selection strategy into its own helper function. I think that will fix `422: Unprocessable entity` errors, which have been happening when subjects were requested with an invalid query. `subjectSelectionStrategy` should only return valid query parameters for each API endpoint.

Test URLs:
https://localhost:8080/?demo=true&project=12268&workflow=17077&subjectSet=92752&env=production
Load subject set 92752 and classify every subject from the first one(priority 129) to the last.

https://localhost:8080/?demo=true&project=12268&workflow=17077&subjectSet=92752&subject=58047423&env=production
Load subject set 92752 and classify every subject from a specified subject (priority 136) to the last.

In demo mode, you can skip quickly through that set by selecting 'Blank page' for every page.

https://localhost:8080?demo=true
Should continue to load I Fancy Cats with random subject selection.

Package:
lib-classifier

Closes #1876.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by shaunanoordin** [comment]
Code read looks good 👍 I'll get to testing on Thursday.

**Comment by eatyourgreens** [comment]
I've rebased this to bring in the latest changes to the project app. If you rebuild the classifier, then you should be able to use the subject picker, from the project home page, and go straight through to classifying your chosen subject. If you classify in demo mode, you should be able to verify that you get all a set's subjects in sequence.
http://local.zooniverse.org:3000/projects/bogden/scarlets-and-blues?env=production&demo=true

**Review by shaunanoordin — APPROVED** [review]
## PR Review

Package: `lib-classifier`
TL;DR: Classifier's subject selection logic has been reorganised, and also now has a clause for "indexed" subject sets.

This PR adds some functionality and cleans up how the Classifier fetches Subjects.

- The "where should I get my Subject from?" logic in the Classifier has been organised into the **subjectSelectionStrategy()** function. _(Personal note: noice 👍 )_
- In addition to the existing methods of selecting Subjects (1. default, use /queued endpoint, 2. workflow is 'grouped', use /queued and specify SSet + WF, 3. Subject Group Viewer, use /grouped endpoint, 4. list of specific Subjects, use /selection endpoint)...
- ...a new method for been added, specifically for **Workflows whose active Subject Set is "indexed".**
  - For this PR, a SSet is "indexed" if it and its WF have the _correct metadata._ (i.e., subject_set.metadata.indexFields = [...], and workflow.grouped = true)
  - ❕  Advanced: this may have an interplay with #2132 , in which the "is this indexed?" question is answered by _querying an [external indexing service.](https://subject-set-search-api.zooniverse.org/)_ 
- ⚠️ Dev note: pay attention to the precedence of the subject selection. If a WF is 'grouped' AND is "indexed", the "is indexed" logic takes priority.

### Testing

Tested on localhost + Chrome + macOS.

- Primary goal is to run several different projects/workflows/subject sets through the Classifier to test that every distinct subject selection clause/scenario works correctly.
- This can be confirmed by checking that the network is making requests to the correct endpoint (e.g. /queued or /selection) with the correct parameters (e.g. ?workflow for standard selection, ?workflow&subject_set for grouped)

Testing results:

1. **Default.** Random subject from `/queued?workflow_id` - [I Fancy Cats](https://localhost:8080/?demo=true) 👍 
2. **Grouped Workflow.** WF has the .grouped attribute, `/queued?workflow_id&subject_set_id` - [Shaun's random test workflow](https://localhost:8080/?demo=true&project=12754&workflow=15652&subjectSet=85773&env=production) 👍 
3. **Subject Group Viewer.** Super specific case, `/grouped?num_columns&num_rows&workflow_id` - [Survos test](https://localhost:8080/?demo=true&env=staging&project=1900&workflow=3412) 👍 
4. **Indexed Subject Set.** WF has subjects on the external Datasette indexing service, which provides a list of 6 Subjects, which then connects to `/selection?ids=...6...` - [Scarlets & Blues](https://localhost:8080/?demo=true&project=12268&workflow=17077&subjectSet=92752&env=production) 👍 
5. **Indexed SSet with specific Subject ID.** As (4), but with an additional specific subject ID - [Scarlets & Blues](https://localhost:8080/?demo=true&project=12268&workflow=17077&subjectSet=92752&subject=58047423&env=production) 👍 
  - Note to self: there are TWO network requests to /selection. The [first](https://www.zooniverse.org/api/subjects/selection?ids=58047423&workflow_id=17077&http_cache=true) is triggered _before_ the requests to Datasette and contains that specified Subject ID, then the [second](https://www.zooniverse.org/api/subjects/selection?ids=58047424%2C58047425%2C58047426%2C58047427%2C58047428%2C58047429&workflow_id=17077&http_cache=true) contains the 6 Subject IDs from Datasette.
  - When I was doing the code read, I thought only the first call would trigger, so huh! I should re-read the code again to fully understand the flow. Still, this functions as expected, so the 👍 remains.

### Status

LGTM! 👍 The new stuff works as expected, and the old stuff ain't broken.

I can think of additional test scenarios (e.g. what if I _drop_ some parameters? Will it florp?) but I don't believe they're necessary to get this PR in. Let's go!

**Comment by eatyourgreens** [comment]
>Note to self: there are TWO network requests to /selection. The first is triggered before the requests to Datasette and contains that specified Subject ID, then the second contains the 6 Subject IDs from Datasette.

The selector needs a subject priority in order to work (get all subjects greater than priority _x_.) The way I've done that is to get the specified subject from Panoptes, get its priority then request 10 IDs from Datasette, greater than that priority. There could well be a way to do this in Datasette with a single query, or maybe with nested queries?
